pkg/ccl/backupccl/backup_test.go

```diff
@@ -1543,6 +1543,61 @@ func TestRestoreRetryProcErr(t *testing.T) {
 	})
 }
 
+func TestRestoreReplanOnLag(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "slow test under stress race")
+
+	retryErrorChan := make(chan error)
+	defer close(retryErrorChan)
+	//Shorten replan frequency setting to reduce test runtime.
+	replanFreq := time.Second
+
+	params := base.TestClusterArgs{}
+	knobs := base.TestingKnobs{
+		BackupRestore: &sql.BackupRestoreTestingKnobs{
+			RunAfterRetryIteration: func(err error) error {
+				retryErrorChan <- err
+				return jobs.MarkAsPermanentJobError(err)
+			},
+			RunAfterRestoreProcDrains: func() {
+				var timer timeutil.Timer
+				defer timer.Stop()
+				// Lag processors with double the duration of replan
+				// frequency such that the timer will be up before
+				// procCompleteCh receives a ping.
+				timer.Reset(replanFreq * 2)
```

Comment:
maybe i'm missing something, but how does this guarantee that a lagging processor will finish more than 2 seconds behind the first processor that completes? I would have expected there to be logic which allows the first processor to complete quickly.

Reply:
The logic in `restore_job` has been updated so that the timer doesn't start until the first processor completes, see this line here: https://github.com/cockroachdb/cockroach/pull/128356/files#diff-d4b5c77a70e36c18a5f100bf7ac6fd728e648db0b6edcb0c1cd7a579a92aedb4R484

In the test we don't need to differentiate between the first processor and the rest because we wouldn't receive pings from the timer channel before the first processor completes, however for subsequent processors the timer will be up before the job gets notified that the processors completed. This also guarantees that we won't replan the job if no processors have completed within 10m, from what I understand we only want to replan if there's an unevenly distributed workload between nodes given by the current dist sql plan, if all nodes are still running after 10m then we should let the job keep running as well.

Reply:
> In the test we don't need to differentiate between the first processor and the rest because we wouldn't receive pings from the timer channel before the first processor completes, however for subsequent processors the timer will be up before the job gets notified that the processors completed

Suppose a restore runs on 3 processors, and each one finishes at exactly 1:00:01 (i.e. 1 am and 1 second), this testing hook would then delay procCompleted msg for _every_ completed proc to 1:00:03 (i.e 1 am and 3 seconds), right? That would imply that all 3 procs finish within the replan window. 

I might just be reading the code incorrectly and we can chat about it on tuesday.

Reply:
>  That would imply that all 3 procs finish within the replan window.

I think we can probably make this test clearer/better, but I think the current test does _work_.  Right now, I think it works by virtue of the fact that processor metas are processed sequentially. So if all processors end at 1:00:01, we will process the first meta and then wait until 1:00:03 to send a message to `progCompletedCh`. _Then_ we will process the second meta and wait until 1:00:00:5 to send the next message to `progCompletedCh`

Reply:
ah, the sequential meta processing is what i was missing. Thanks!